### PR TITLE
Add Stormworks

### PR DESCRIPTION
**Pull request by JonathanArendt**
After much testing, added full support for Stormworks Dedicated Server for AMP on Windows. The server is able to start, stop, update. Added all of the configuration options to the config manifest to add support for editing the server config file as little as possible, the only outstanding features are Linus support. And the ability to manage whitelist, blacklist, admin, ect, within AMP, this is in part to a lack of the ability to import a list with the same ID, and in part due to how the game's config is.

### CODE REVIEW

**Comment by IceOfWraith** [comment]
It looks like @Greelan added Linux support and you've done some extra clean up also. Is this ready for review? I'll knock it out this weekend if so.

**Comment by JonathanArendt** [comment]
@IceOfWraith I have changes I have not pushed yet, one to fix a bug with whitelist being enabled by default, and I need to test what, AMP will do with muliple tags with the same index, (Idealy, amp should set at least one admin, so then the other whitelist/blacklists, can be handeled in game. And if that doesn't work I will had type Hidden boxes for instructions on setting up permissions,. All that said, I'm going to test a few more things, and then I'll let you know when It's good. Should have it ready by this weekend.

**Comment by IceOfWraith** [comment]
No rush! I just didn't want it waiting on me. :)

**Comment by Greelan** [comment]
I still need to do further review too

**Comment by JonathanArendt** [comment]
All good, I was just going over some spelling ;), and other small things. Everything should be good now, I added a text field for admin, it sets the first admin in the list on the xml and that's it. As added in the notes, almost everything else can be controlled from in game.

**Comment by JonathanArendt** [comment]
At this time, I have everything set up and ready for review. I would ask that you read over some of the "documentation" within the settings menus for any spelling issues and anything else that may need to be reworded or clarified.

**Comment by JonathanArendt** [comment]
@IceOfWraith I'm going to work on typing up some information for you add in the docs for Additional Information relating to game permissions, in game commands, and how to setup Whitelists.

**Comment by JonathanArendt** [comment]
@IceOfWraith I just pushed stormworksguide.md, this is really just a rough draft for the discourse server, I read that discourse uses Markdown so sorry if that's extra work for you, lol. Feel free to rewrite, reformat, add, remove, ect, to this. It's simply the information that I felt would feel crowded on the settings menu. Let me know if I need to add any of the information to the settings menu.

**Comment by JonathanArendt** [comment]
@Greelan I'm not sure what steps are needed for Linux support so I need to add any additional settings or information for the guide, feel free to add to it or let me know.

**Comment by Greelan** [comment]
It just needs Wine, desirably Wine 8. So either use docker or install wine 8 on the host. The docker policy box has that info.

**Comment by Greelan** [comment]
> I read that discourse uses Markdown so sorry if that's extra work for you, lol. 

.md on GH is markdown too lol

**Comment by JonathanArendt** [comment]
Lol, I know that it's markdown, I'm not sure if markdown is actually what's used on discourse, lol. I know some blogs use other languages, lol.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by Greelan** [comment]
Yes, discourse uses markdown

**Comment by JonathanArendt** [comment]
Have either of you done any further review? I just got off work and wanted to check if there was any outstanding issues that need to be worked on?

**Comment by IceOfWraith** [comment]
I think I'm just holding for Greelan to finish his reviewing.

**Comment by JonathanArendt** [comment]
Sounds good, I know it's a holiday weekend, (at least for us in the US (don't know where you all are form)) so don't know if that means there will be more work done on this or less, lol.  Either way, it's all good, I'll be here if there's any changes to be made this weekend.

**Comment by IceOfWraith** [comment]
I'm in the US so definitely spending time on AMP. When he says he's good I'll go through and do any followup.

**Comment by JonathanArendt** [comment]
Sounds good to me, thanks Ice!

________________________________

**Comment by Greelan** [comment]
I will get to it this week

**Comment by Greelan** [comment]
@JonathanArendt , I've committed some clean-ups to the config manifest. I still need to be do testing on the settings to confirm they write correctly.

Couple of things:
- I suggest that the Start Island setting list be re-ordered in alphabetical order, to help people find islands
- If the Day & Night Length setting can only be in increments of 10, is the MinValue correct at 1? Would it be better for this to be an enum, with the valid settings specified (10, 20, 30 etc)?

I haven't looked at your guide in detail. On a quick skim I thought it could be with some tidy up (typos etc) and also some bits should be re-written, as it wasn't clear what they actually meant. Also, it shouldn't be in the PR.

**Comment by Greelan** [comment]
BTW, the admins setting does not appear to write anything to the config file

**Comment by JonathanArendt** [comment]
So for the day night setting, a minimum of one is correct, since you can't have zero time, 1 is the minimum setting and after that you it's even 10s. (not me, that's the devs, I could go on, but just look at how the console works, lol)

**Comment by JonathanArendt** [comment]
As for the list, I used the ordering on the guide, It's listed based on the island that the player starts on (and that includes dlc bases on the dlc maps). I used the same ordering on the server setup guide, but i could change the order of it, if that makes it more clear. I wonder if I do that if I should add notes as to the location of the island or the dlc needed, but that might make it too cluttered.

**Comment by Greelan** [comment]
> So for the day night setting, a minimum of one is correct, since you can't have zero time, 1 is the minimum setting and after that you it's even 10s. (not me, that's the devs, I could go on, but just look at how the console works, lol)

OK, but how about enums then? 1, 10, 20, etc

**Comment by JonathanArendt** [comment]
That works, I wasn't sure which would be more clear, I'll up date that now

**Comment by Greelan** [comment]
> As for the list, I used the ordering on the guide, It's listed based on the island that the player starts on (and that includes dlc bases on the dlc maps). I used the same ordering on the server setup guide, but i could change the order of it, if that makes it more clear. I wonder if I do that if I should add notes as to the location of the island or the dlc needed, but that might make it too cluttered.

Yeah, I think you need to distinguish what is DLC and what is stock. Eg for OpenRCT2 I add the prefixes `RCT1:` and `RCT2:`.

Regardless of the order that the guide lists it, it is easier for someone to find a specific island via alphabetical order, given the list is quite long - you might have it in blocks: first stock islands in alpha order, then DLC islands in alpha order

**Comment by JonathanArendt** [comment]
> 

There's no way to had text delimiters in the enum list correct? like a non selected vaule to sepereate the list?

**Comment by Greelan** [comment]
Afraid not

**Comment by Greelan** [comment]
@IceOfWraith , over to you xD

**Comment by JonathanArendt** [comment]
@IceOfWraith I will be busy this weekend, but I will probably be availible for questions, and then I'll be free on monday if needbe.

**Comment by IceOfWraith** [comment]
I'm just diving into this one now. I do have a question about the note here:

`These settings can also be changed in-game if "Enable Settings Menu" under "Stormworks - World Creation" is set.`

If a user sets a setting in-game, but AMP has a setting that differs, does the setting get reverted when they restart the server since AMP isn't able to detect changes made to a settings file elsewhere?

**Comment by JonathanArendt** [comment]
> 

So No Amp does not override the settings, they are only the Default setting, and only matter for world creation, the reason I have them in a seperate group is because those settings can be changed in game if the settings are enabled, however the World creation settings cannot be changed after creation.

The only settings that affect the server after creation, is the Server Settings.

**Comment by IceOfWraith** [comment]
I don't own the game to see it myself, so that's why I have to rely on you to provide this info. I assume that means on first time in the game you can choose the defaults, and after that it sticks with whatever was chosen. And none of the settings under Server Settings are editable in game.

**Comment by JonathanArendt** [comment]
Correct, Server/General settings (aside from Admin, Auth, and Blacklist) are not editable in game, but they do change the server when the server is started, 
World creation settings can only be set when the save is generated, you either have to specify a new save, or delete the existing save to update.
The Gameplay settings can be changed ingame if the Settings Menu is enabled for admins, otherwise they act as World Creation settings.

**Comment by JonathanArendt** [comment]
If you have questions and want to discuses them, we can hop into Discord Voice Channel and I can help facilitate testing with the game, if need be.

**Comment by IceOfWraith** [comment]
I don't think that's necessary so far. I see that it starts up normally in Windows and Linux. I'm going through and reviewing the settings for spelling/grammar. It looks like you did a good job at first glance. I do note some spelling mistakes, but I'll provide that all when I'm done. :)

**Comment by JonathanArendt** [comment]
Sounds good, Note that the extra steps guide needs the correct url before release, I started the .md file for you, but it may need to be seriously overhauled. And it needs to be removed from the PR before pushing the changes.

**Comment by IceOfWraith** [comment]
Definitely, the config file update stage will also need updated. (noting for myself) Lol

**Comment by JonathanArendt** [comment]
I'm not good at technical writing, but I can help clarify the points I'm trying to get across. There are some side notes that you might want to add, The stormworks server is a defiantly a weird one. I tried to be clear on the Notes in settings menus, but idk if it could be made more clear.

**Comment by JonathanArendt** [comment]
Such as this [Setting up Loopback Adapter on same network.](https://steamcommunity.com/sharedfiles/filedetails/?id=2433754130&searchtext=advanced+dedicated#:~:text=Installing%20Microsoft%20Loopback%20Adapter)

**Comment by IceOfWraith** [comment]
Thanks, I hate it.

**Comment by JonathanArendt** [comment]
Hate what??
The game?

**Comment by IceOfWraith** [comment]
The fact you need the loopback adapter installed to see the server on the same machine. Lol

**Comment by JonathanArendt** [comment]
Yeah same, It's weird, I was working on testing the RCT2 for Greelan and it was causing it to show up twice, lol.

**Comment by JonathanArendt** [comment]
Hey, do you want me to go throught and add notes for review on the url's and stuff so we don't forget to change any of them.

**Comment by IceOfWraith** [comment]
Way ahead of you. Give me a few more minutes and you should have your review ready. :)

File: stormworksconfig.json [hunk 1]
```diff
@@ -0,0 +1,917 @@
...
+      "0": "None",
+      "1": "Low (default)",
+      "2": "Medium",
+      "3": "High"
+    }
+  },
+  {
+    "DisplayName": "World Creation Settings",
+    "Category": "Stormworks - World Creation",
+    "Description": "These settings are for world creation. They may not be able to be changed after the world is created. If needed, delete  the \"working_server\" and \"saves\" directories to generate a new world.",
```
Comment (IceOfWraith) [inline stormworksconfig.json:184]: `They may not be changed`

File: stormworksconfig.json [hunk 2]
```diff
@@ -0,0 +1,917 @@
...
+      "data/tiles/island_30_playerbase_boat.xml": "Custom: Boat Base Large",
+      "data/tiles/island_32_playerbase_heli.xml": "Custom: Heli Base (structure workbench only)",
+      "data/tiles/oil_rig_playerbase.xml": "Custom: Oil Rig Base (structure workbench only)",
+      "data/tiles/island_29_playerbase_submarine.xml": "Custom: Submarine Base"
+    }
+  },
+  {
+    "DisplayName": "Starting Currency",
+    "Category": "Stormworks - World Creation",
+    "Description": "Sets the ammount of money the players start out with at world creation.",
```
[the remark this answers is not in the document]
Comment (IceOfWraith) [reply stormworksconfig.json:256]: `players start with`

**Comment by JonathanArendt** [comment]
Lol sounds good, I'm used to small changes and then pushing one at a time, i thought you were just making a list, lol.

File: stormworksconfig.json [hunk 3]
```diff
@@ -0,0 +1,917 @@
...
+    "DefaultValue": "false",
+    "EnumValues": {
+      "False": "false",
+      "True": "true"
+    }
+  },
+  {
+    "DisplayName": "Gameplay Settings",
+    "Category": "Stormworks - Gameplay",
+    "Description": "These settings are for configuring gameplay options for a world. Note that changing them may not update a save game. These settings can also be changed in-game if \"Enable Settings Menu\" under \"Stormworks - World Creation\" is set. If needed, delete  the \"working_server\" and \"saves\" directories to generate a new world, or try to edit \"scene.xml\" under each of these folders (untested).",
```
Comment (IceOfWraith) [inline stormworksconfig.json:335]: `delete the` (extra space)

**Comment by JonathanArendt** [comment]
Oh, one of the biggest things I hate about this game is the console output, not that it's neccasarily bad, cause on CMD it just clears the screan each time, it doesn't spam it, but on amp it just is spam, It would be noice if amp had IO mode that recgonises this and just updates a graft or something, but that's up to Mike to implement is that right?

File: stormworksguide.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+Admins are able to run the in game commands that are listed below, as well as change the settings in the Settings Menu, if enabled in the Stormworks - World settings. 
+
+### Authorized
+
+ In game, players have to be authorized in order to use the workbench, spawn in creatations, ect. 
+
+### Blacklist
+
+Simply a ban list, players can be banned in game, however, you have to remove their id from the ban list, similar to adding a player to the whitelist, however remove the `<id value="TheirID" />`.
```
Comment (IceOfWraith) [inline stormworksguide.md:21]: `Simply a ban list. Players`

Can you provide details of which file has these values in it that will need removed?

File: stormworksguide.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+Example:
+```xml
+<whitelist>
+    <id value="Steam64ID"/>
+</whitelist>
+```
+
+## Commands
+
+The following commands can be used in game to control player permissions and to save the game. Note that the game's console does not have inputs, so commands must be ran in game. This means that there is no way of saving the game aside from running the in game command. You can press the enter key to open the chat box and run the commands.
```
[the remark this answers is not in the document]
Comment (IceOfWraith) [reply stormworksguide.md:37]: Does the game not have any type of autosave? That's scary if so!

File: stormworksguide.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
+# Stormworks Setup Guide
+
+## Linux Support
+
+The server requires Wine to run on Linux, Preferably Wine 8. Using a container avoids the need to install this dependency on the host, otherwise,  Wine is required to be installed on the host system to run the server. 
+
+## Player Permissions
+
+There are 4 settings for player permissions. Admins, Authorized, Blacklist, and Whitelist.
```
Comment (IceOfWraith) [inline stormworksguide.md:9]: Perhaps, note that you have to use the commands at the bottom to adjust these in-game? The way it reads now, I was confused what I'd do with these until I finished the whole guide.

**Review by IceOfWraith — CHANGES_REQUESTED** [review]
Let me start by saying, great job! This is a very thorough template. The main issues noted are grammar and spelling. I think I've grabbed all of them in the review. There are a few suggestions for the guide portion as well.

I'll create an article as soon as the guide portion is done so you can get a link to use for the other edits I mentioned.

Please respond with any changes or reasons why I'm wrong (which happens a lot), and we can get this merged. You can close the individual review comments as you make changes.

Again, great work on this!

**Comment by IceOfWraith** [comment]
> Oh, one of the biggest things I hate about this game is the console output, not that it's neccasarily bad, cause on CMD it just clears the screan each time, it doesn't spam it, but on amp it just is spam, It would be noice if amp had IO mode that recgonises this and just updates a graft or something, but that's up to Mike to implement is that right?

Yupp, nothing can be done from our end on this front. He does intend to revamp the console though. So hopefully it comes out much smoother. It's been a pain point for generic forever.

**Comment by JonathanArendt** [comment]
@IceOfWraith  I have resolved all of the current spelling changes, and clairied the .md file.

**Comment by IceOfWraith** [comment]
Very good! I'll look again tomorrow and let you know what I see.

**Comment by IceOfWraith** [comment]
It looks like there's still some items that aren't resolved. Check the conversation above and you'll see the review comments. You might need to unhide the messages since GitHub likes to shrink it.

**Comment by JonathanArendt** [comment]
> It looks like there's still some items that aren't resolved. Check the conversation above and you'll see the review comments. You might need to unhide the messages since GitHub likes to shrink it.
The only conversations that i did not resolve where one's that need to be verified because I changed the text.
or the urls

**Comment by JonathanArendt** [comment]
This is my first PR, so I'm not sure if there's something else that I need to do, I didn't close the URLs because they need to be changed still, and the one's where I reworded something, I didn't close it but requesed review on that item so you know where to look.

File: stormworks.kvp [hunk 7]
```diff
@@ -0,0 +1,99 @@
+Meta.DisplayName=Stormworks
+Meta.Description=Stormworks Dedicated Server
+Meta.OS=Windows, Linux
+Meta.AarchSupport=NotSupported
+Meta.Arch=x86_64
+Meta.Author=SauranTheWhite,Greelan
+Meta.URL=https://store.steampowered.com/app/573090/Stormworks_Build_and_Rescue/
+Meta.DisplayImageSource=steam:573090
+Meta.EndpointURIFormat=steam://connect/{0}
```
[the remark this answers is not in the document]
Comment (IceOfWraith) [reply stormworks.kvp:9]: @JonathanArendt This is the first one that I don't see addressed. It either needs changed, or explanation why it wasn't changed. There are sometimes valid reasons to not do these suggestions, like if a game doesn't support a feature.

File: stormworksconfig.json [hunk 8]
```diff
@@ -0,0 +1,917 @@
...
+    "Description": "Sets the physics resolution.",
+    "Keywords": "physics,timestep,resolution,physicstimestep,physics_timestep",
+    "FieldName": "physics_timestep",
+    "InputType": "enum",
+    "IsFlagArgument": false,
+    "ParamFieldName": "/server_data/@physics_timestep",
+    "IncludeInCommandLine": false,
+    "DefaultValue": "0",
+    "EnumValues": {
+      "0": "Auto (default)",
```
[the remark this answers is not in the document]
Comment (IceOfWraith) [reply stormworksconfig.json:141]: @JonathanArendt Here's another one.

File: stormworks.kvp (hunk 7 above)
[the remark this answers is not in the document]
Comment (JonathanArendt) [reply stormworks.kvp:9]: That's what I was saying, Steam connect did not work, And as far as I know, the only way to connect is in game, there is no direct connect option in game.

**Comment by IceOfWraith** [comment]
No need to worry. I don't mind helping you through the process. I struggled like crazy my first time. I'm pretty sure I even had to close and open a new one because I botched it so bad. Lol

I tagged you in a couple of comments, I'm not sure if that will show easily for you.

File: stormworksconfig.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (JonathanArendt) [reply stormworksconfig.json:141]: I pushed changes to include the number's in parrenthese, is that acceptable?

**Comment by JonathanArendt** [comment]
yeah, so I did address each issue in one way or an other, And i posted comments on each one, don't know if that's how I'm supposed to, but I leaft them open because I felt like they needed discussion for resolution.\

**Comment by JonathanArendt** [comment]
And i responded to you're comment's that you just posted for clearification as well.

**Comment by IceOfWraith** [comment]
Hmm... I don't see your responses for some reason. Maybe I'm looking in the wrong place. x)

**Comment by JonathanArendt** [comment]
Ohh

**Comment by JonathanArendt** [comment]
Oh, I don't know, they all say pending?

**Comment by IceOfWraith** [comment]
Ahhh okay, I think you're posting a review also. You should be able to respond directly to my comments up above here. There should be a Reply box under each one.

**Comment by JonathanArendt** [comment]
image

**Comment by IceOfWraith** [comment]
Well that's new... Let me see what's up.

**Comment by IceOfWraith** [comment]
WOAH! There they all are. Haha

**Comment by JonathanArendt** [comment]
yeah, I don't know, but now you should see everyting that I did last nigh,

**Comment by JonathanArendt** [comment]
I had to post my reveiw, that's stupid, sorry, didn't know I had to do that

**Comment by JonathanArendt** [comment]
image

**Comment by JonathanArendt** [comment]
https://github.com/orgs/community/discussions/6893

**Comment by JonathanArendt** [comment]
Apparently it's a bug/issue, idk. I've only used github for repos and personal projects with frinds, not like this.

**Comment by IceOfWraith** [comment]
Oof. Well that's good to know! I haven't run into that yet on any of my reviews. (surprisingly)

Or maybe I have and Greelan just knew what I was trying to say. Haha

**Comment by JonathanArendt** [comment]
Are you and greleen the only ones that make templates?

**Comment by IceOfWraith** [comment]
The only regulars, yeah. We've had others submit some awesome things though. TmodLoader and mods for 7D2D were big ones others provided.

**Comment by JonathanArendt** [comment]
Well, I don't mind helping, I enjoy it, It's an interesting side of coding.

**Comment by JonathanArendt** [comment]
> Oof. Well that's good to know! I haven't run into that yet on any of my reviews. (surprisingly)
> 
> Or maybe I have and Greelan just knew what I was trying to say. Haha

It's almost like I have to respond and then post my response at once, so he may know that. Once I hit submit review, everything showed up at once, i don't know. I may have done it wrong thou.

**Comment by IceOfWraith** [comment]
It's always appreciated! You seem to have grabbed the best parts of other templates and put them into this one which a lot of others have failed to do on their first try. So great job with it!

It's getting late here so I'll finish reviewing tomorrow, but I'm glad to see you actually had replied and it was just a glitch.

**Comment by JonathanArendt** [comment]
Sounds good, The big one is the MD, it got a lot of rewrites, everything else should be resolved, the two that you pinged me on you may need to check if it's acceptable or advise on steps to be taken.

File: stormworksconfig.json [hunk 9]
```diff
@@ -0,0 +1,917 @@
...
+    "ParamFieldName": "/server_data/admins/id[1]/@value",
+    "IncludeInCommandLine": false,
+    "DefaultValue": "",
+    "Placeholder":"9087654321234567",
+    "EnumValues": {}
+  },
+  {
+    "DisplayName": "Player Permissions",
+    "Category": "Stormworks - General",
+    "Description": "Player permissions (and whitelisted players) can be set using in-game commands. See the [Stormworks Guide](https://discourse.cubecoders.com/docs?category=21) for further information.",
```
Comment (JonathanArendt) [inline stormworksconfig.json:32]: Needs updated before merge.

File: stormworksconfig.json [hunk 10]
```diff
@@ -0,0 +1,917 @@
...
+    "IsFlagArgument": false,
+    "ParamFieldName": "",
+    "IncludeInCommandLine": false,
+    "DefaultValue": "",
+    "EnumValues": {}
+  },
+  {
+    "DisplayName": "Primary Admin",
+    "Category": "Stormworks - General",
+    "Description": "Sets the [Steam64 ID](https://steamidfinder.com/) of the primary (first) admin on the server. Other admins (as well as authorized users and banned users) can be added in-game. See the [Stormworks Guide](https://discourse.cubecoders.com/docs?category=21) for further information.",
```
Comment (JonathanArendt) [inline stormworksconfig.json:18]: Needs updated before merge.

**Comment by JonathanArendt** [comment]
Do you want to remove the guide from the PR as well?

**Comment by IceOfWraith** [comment]
Oh yeah, go ahead and remove that also.

**Comment by JonathanArendt** [comment]
All done!

**Review by IceOfWraith — APPROVED** [review]
Great work! Thank you for the submission!

**Pull request merged**